catalyst/callbacks/metric.py

```diff
@@ -1,78 +1,300 @@
 # @TODO: add metric aggregation, etc callback
+from typing import Dict, Iterable, Optional, Tuple, Union
+from abc import ABC, abstractmethod
+
+import torch
+
 from catalyst.core.callback import Callback, CallbackNode, CallbackOrder
 from catalyst.core.runner import IRunner
-from catalyst.metrics._metric import ICallbackBatchMetric, ICallbackLoaderMetric
+from catalyst.metrics import (
+    ICallbackBatchMetric,
+    ICallbackLoaderMetric,
+    IMetric,
+)
+
+TORCH_BOOL = torch.bool if torch.__version__ > "1.1.0" else torch.ByteTensor
 
 
-class IMetricCallback(Callback):
+class IMetricCallback(Callback, ABC):
     """Metric callback interface, abstraction over metric step."""
 
-    pass
+    @abstractmethod
+    def on_loader_start(self, runner: "IRunner") -> None:
+        """
+        On loader start action
+
+        Args:
+            runner: current runner
+        """
+        pass
 
+    @abstractmethod
+    def on_batch_end(self, runner: "IRunner") -> None:
+        """
+        On batch end action
+
+        Args:
+            runner: current runner
+        """
+        pass
+
+    @abstractmethod
+    def on_loader_end(self, runner: "IRunner") -> None:
+        """
+        On loader end action
 
-# @TODO: add KV support for input/output
-class BatchMetricCallback(IMetricCallback):
-    """@TODO: docs."""
+        Args:
+            runner: current runner
+        """
+        pass
 
+
+class MetricCallback(IMetricCallback):
     def __init__(
         self,
-        metric: ICallbackBatchMetric,
-        input_key: str,
-        target_key: str,
-        log_on_batch: bool = True,
+        metric: Union[ICallbackBatchMetric, ICallbackLoaderMetric],
+        input_key: Union[str, Iterable[str], Dict[str, str]],
+        target_key: Union[str, Iterable[str], Dict[str, str]],
     ):
-        """@TODO: docs."""
+        """
+        Init MetricCallback
+
+        Args:
+            metric: metric to calculate in callback
+            input_key: keys of tensors that should be used as inputs in metric calculation
+            target_key: keys of tensors that should be used as targets in metric calculation
+        """
         super().__init__(order=CallbackOrder.metric, node=CallbackNode.all)
-        assert isinstance(metric, ICallbackBatchMetric)
         self.metric = metric
+        self._validate_metric(metric=metric)
+        self._metric_update_method = self.metric.update
```

Comment:
why do we need it?

Reply:
I see, that u are using `self.metric.update` and `self.metric.update_key_value`, but it would be great if you could explain the case also :)